text/0000-generic-dependent-consts.md

```diff
@@ -0,0 +1,291 @@
+- Feature Name: generic_dependent_consts
+- Start Date: 2015-03-28
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Allow consts declared in functions to depend on function type parameters. Since
+this raises some issues in match and type checks, use an conservative approach
+that allows only minimal use of such consts when a constant expression is
+required.
+
+# Motivation
+
+Consider the following line of code in a non-generic context:
+
+```rust
+let a: [u8; T::N + T::N] = [0u8; 2*T::N];
```

Comment:
Technically, if it's part of the signature, it will be detected before instantiating the function's body with the given type parameters (i.e. "monomorphization").
If it's not part of the signature, then the operation would generate a runtime panic if it were operating on runtime values.

The way I see it, such compiler errors are "early warnings" for runtime errors - you can still get a program that type-checks if you define the result of all operations which could fail.

It is actually possible to codegen such a function to panic at runtime instead of emitting a compile-time error.
It could also trigger a lint, so you can get an error for functions which unconditionally panic, *if you want that* (could be an error by default, too).